fairscale/optim/oss.py

```diff
@@ -382,16 +387,95 @@ def _broadcast_state_dict(self) -> None:
-            if i == self.rank:
+        for rank, partition in enumerate(self.partition_parameters()):
+            if rank == self.rank:
                 continue
 
             for p in partition:
                 for t in p["params"]:
                     t.grad = None
+
+    @staticmethod
+    def get_global_rank(group: Any, rank: int) -> int:
+        if group is dist.group.WORLD:
+            return rank
+        else:
+            global_rank = dist.distributed_c10d._get_global_rank(group, rank)  # type: ignore
+        return global_rank
+
+    @staticmethod
+    def _consume_async_work(queue: List[Any]) -> None:
+        _ = list(map(lambda x: x.wait(), queue))
+
+    @staticmethod
+    def _broadcast_params(
+        buffers: List[torch.Tensor], per_rank_params: List[List[Parameter]], group: Any, self_rank: int
+    ) -> None:
+        """Helper function to broadcast all the parameters from a given device
+        """
+        buffer_size = buffers[0].numel()
+        restore_require_grad = []
+        bucket_requests = []
+        requests = []
+
+        # Bucket and issue all the async calls
+        for (rank, params), buffer in zip(enumerate(per_rank_params), buffers):  # all the params sorted per rank
+            if len(params) == 0:
+                continue
+
+            global_rank = OSS.get_global_rank(group, rank)
+
+            # Copy small parameters into per-GPU buffers
+            i_bucketed = 0  # the number of tensors packed in the buffer
+            offset = 0
+
+            while i_bucketed < len(params) and offset + params[i_bucketed].numel() < buffer_size:
+                end = offset + params[i_bucketed].numel()
+                if rank == self_rank:
+                    buffer[offset:end].copy_(params[i_bucketed].data.view(-1))  # type: ignore
+                offset = end
+                i_bucketed += 1
+
+            if i_bucketed > 0:
+                future = dist.broadcast(tensor=buffer, src=global_rank, group=group, async_op=True)
+                if rank != self_rank:
+                    # This request will need to be unrolled
+                    bucket_requests.append((future, rank))
+
+            # Directly broadcast the rest
+            for param in params[i_bucketed:]:
+                # NOTE: Broadcast is in-place and not differentiable
+                # Gloo will assert on this operation for any tensor that requires grad.
+                # We save and restore the grad requirement state to work around that, in our case
+                # the grad is only useful on the source rank.
+                if param.requires_grad:
+                    restore_require_grad.append(param)
+                    param.requires_grad = False
+
+                requests.append(dist.broadcast(tensor=param, src=global_rank, group=group, async_op=True))
+
+        # Unroll the initial packed small parameters
+        for gate, rank in bucket_requests:
```

Comment:
earlier we are calling this a future and here we call it a gate. might be simpler to use consistent names.

Reply:
when unrolling the requests this acts as a gate (these "futures" are not really futures anyway, we cannot act on them), so I thought that was clearer to the reader. Apparently not, changing that